cdap-api/src/main/java/co/cask/cdap/api/workflow/WorkflowSpecification.java

```diff
@@ -55,381 +34,81 @@
  *  <pre>
  *    <code>
  *      {@literal @}Override
- *      public WorkflowSpecification configure() {
- *        return WorkflowSpecification.Builder.with()
- *        .setName("PurchaseHistoryWorkflow")
- *        .setDescription("PurchaseHistoryWorkflow description")
- *        .onlyWith(new PurchaseHistoryBuilder())
- *        .addSchedule(new Schedule("DailySchedule", "Run every day at 4:00 A.M.", "0 4 * * *",
- *                     Schedule.Action.START))
- *        .build();
+ *      public void configure() {
+ *        setName("PurchaseHistoryWorkflow");
+ *        setDescription("PurchaseHistoryWorkflow description");
+ *        addMapReduce("PurchaseHistoryBuilder");
+ *        addSchedule("DailySchedule");
  *      }
  *    </code>
  *  </pre>
  *
  * See the Purchase example application.
  */
-public interface WorkflowSpecification extends SchedulableProgramSpecification {
-
-  List<WorkflowActionSpecification> getActions();
-
-  Map<String, MapReduceSpecification> getMapReduce();
-
-  Map<String, SparkSpecification> getSparks();
-
-  /**
-   * Builder for adding the first action to the workflow.
-   * @param <T> Type of the next builder object.
-   */
-  public interface FirstAction<T> {
-
-    MoreAction<T> startWith(WorkflowAction action);
-
-    MoreAction<T> startWith(MapReduce mapReduce);
-
-    MoreAction<T> startWith(Spark spark);
-
-    T onlyWith(WorkflowAction action);
-
-    T onlyWith(MapReduce mapReduce);
-
-    T onlyWith(Spark spark);
+public final class WorkflowSpecification implements ProgramSpecification, PropertyProvider {
+  private final String className;
+  private final String name;
+  private final String description;
+  private final Map<String, WorkflowActionSpecification> customActionMap;
+  private final List<ProgramNameTypeInfo> actions;
```

Comment:
actions contains list of programs which Workflow is supposed to execute sequentially. These programs can be either MapReduce, Spark or Custom_Action. Specifications for MapReduce and Sparks are stored at Application level, however customAction map stores custom action specifications. I think its better to keep them separate. Please let me know your thoughts.